src/toast/io/compression.py

```diff
@@ -0,0 +1,274 @@
+
+    """
+    log = Logger.get()
+
+    if comp_params is None:
+        comp_params = {"type": "none"}
+    if "type" not in comp_params:
+        raise ValueError("comp_params must contain the 'type' key")
+
+    comp_params["detectors"] = detdata.detectors
+    n_det = len(detdata.detectors)
+    comp_params["det_shape"] = detdata.detector_shape
+    comp_params["dtype"] = detdata.dtype
+    comp_params["units"] = detdata.units
+
+    msg = f"Compressing detector data {detdata} with {comp_params}"
+    log.verbose(msg)
+
+    det_stride = 1
+    for ds in detdata.detector_shape:
+        det_stride *= ds
+
+    if comp_params["type"] == "none":
+        det_stride = 1
+        for ds in detdata.detector_shape:
+            det_stride *= ds
+        det_bytes = det_stride * detdata.dtype.itemsize
+        comp_ranges = [(x * det_bytes, (x + 1) * det_bytes) for x in range(n_det)]
+        comp_bytes = np.array(detdata.flatdata.view(dtype=np.uint8), dtype=np.uint8)
+        return (
+            comp_bytes,
+            comp_ranges,
+            comp_params,
+        )
+
+    elif comp_params["type"] == "gzip":
+        comp_bytes = AlignedU8()
+        off = 0
+        if "level" in comp_params:
+            comp_level = comp_params["level"]
+        else:
+            comp_level = 9
+            comp_params["level"] = comp_level
+        start_bytes = list()
+        for idet, det in enumerate(detdata.detectors):
+            start_bytes.append(off)
+            dbytes = gzip.compress(
+                detdata[det, :].tobytes(), compresslevel=comp_level, mtime=0
+            )
+            nb = len(dbytes)
+            comp_bytes.resize(off + nb)
+            comp_bytes[off : off + nb] = dbytes
+            off += nb
+        comp_ranges = list()
+        last = None
+        for idet in range(n_det):
+            if idet == n_det - 1:
+                cur = (last[1], len(comp_bytes))
+            else:
+                cur = (start_bytes[idet], start_bytes[idet + 1])
+            comp_ranges.append(cur)
+            last = cur
+        return (comp_bytes.array(), comp_ranges, comp_params)
+
+    elif comp_params["type"] == "flac":
+        ftypes = [np.dtype(np.float32), np.dtype(np.float64)]
+        itypes = [np.dtype(np.int32), np.dtype(np.int64)]
+
+        # Check parameters
+        if "level" in comp_params:
+            comp_level = comp_params["level"]
+        else:
+            comp_level = 5
+            comp_params["level"] = comp_level
+        if "quanta" in comp_params:
+            quanta = comp_params["quanta"]
+            if detdata.dtype in itypes:
+                raise RuntimeError(
+                    "Cannot specify quanta for FLAC compression of integers"
+                )
+        else:
+            quanta = None
+            comp_params["quanta"] = quanta
+
+        if detdata.dtype in ftypes:
+            (
+                comp_bytes,
+                comp_ranges,
```

Comment:
I would be interested to see how polynomial detrending would reduce the dynamical range of the data. Also, the data we simulate probably could do with more digitization noise to improve the compression ratio.

Reply:
The FLAC algorithm does do [modeling of the signal](https://xiph.org/flac/format.html#prediction) to reduce the residuals that need to be encoded.  However I believe this is done on spans of the blocksize number of samples (e.g. 4096 or 8192 samples).  Certainly doing a more complicated fit would be possible, in which case we should replace the gain / offset model with a set of coefficients for each detector.  Also note by default we map the full range of data to +/- 2^30, but one can set the quanta for all detectors to a constant from the commandline / config for the SaveHDF5 operator.